### ISSUE TITLE
Question about the weight of two losses

### ISSUE THREAD

**Issue by wincle** [FlyingGoblin/CartoonGAN#3]
Hi , first of all , thanks for your share!
   I want to ask about that the weight of two losses, adv loss and content loss, how do you set the balance of them? I found in your code, both of them you mul 10, but in the paper ,only the content loss mul 10.Could you tell me why make this change or where am I wrong?
  Thanks again!

**Comment by wincle** [comment]
上面写英文是为了国际友人也能理解，这里中文再阐释下，上面可能说的可能不太清楚。
我看论文里作者设置content loss的权重为10，而 adv loss没有乘以任何权重，但是我看你的代码里，都乘以了10（或者是我对代码理解错了）。其实我自己在用其他作者提供的CartoonGan训练代码训练时，按照作者那种10:1的loss设置方式，发现训练了很多轮迭代，但是丝毫没有卡通风格出现，所以怀疑论文里作者出现了笔误，而且在我理解，如果content loss比重过大，那么不管训练多少轮迭代，也只是让输入输出竭力相似，所以猜测应该是adv loss为10，content loss为1，我用这种比重试验了下，确实效果要好很多，所以想和你验证一下，你是否遇到类似的问题？谢谢。

**Comment by FlyingGoblin** [comment]
the code of adv loss for the generator is in the line 182~188 of cartoon_gan_model.lua, 0-0 I don't think I mul 10
`-- GAN loss
  local output = netD_source.output -- [hack] forward was already executed in fDx, so save computation netD_source:forward(fake_B) ---
  local errGAN = self.criterionGAN:forward(output, self.real_label)
  local df_do = self.criterionGAN:backward(output, self.real_label)
  local df_d_GAN = netD_source:updateGradInput(fake_target, df_do) ---:narrow(2,fake_AB:size(2)-output_nc+1, output_nc)
  netG_source:forward(real_source)
  netG_source:backward(real_source, df_d_GAN)`

**Comment by FlyingGoblin** [comment]
the code of adv loss for the generator is in the line 182~188 of cartoon_gan_model.lua, 0-0 I don't think I mul 10
-- GAN loss

  local output = netD_source.output -- [hack] forward was already executed in fDx, so save computation netD_source:forward(fake_B) ---

  local errGAN = self.criterionGAN:forward(output, self.real_label)

  local df_do = self.criterionGAN:backward(output, self.real_label)

  local df_d_GAN = netD_source:updateGradInput(fake_target, df_do) ---:narrow(2,fake_AB:size(2)-output_nc+1, output_nc)

  netG_source:forward(real_source)

  netG_source:backward(real_source, df_d_GAN)

**Comment by FlyingGoblin** [comment]
content loss的权重确实不能太大，否则就会出现风格没有办法迁移的问题。
![image](https://user-images.githubusercontent.com/8320622/46124573-94408600-c257-11e8-91f9-3ca7b71c2fd6.png)
其实我在论文提交时的补充材料里讨论过权重的影响，但是组委会好像后来并没有上传补充材料┓( ´∀` )┏

**Comment by wincle** [comment]
哦哦~，谢谢，明白了。
原来您才是原作啊，居然Star这么少，可能是用lua的人不多吧，我是用了另一个人的python-tensorflow的代码训练的，既然原作在此，要好好研究一下您的代码了。

**Comment by wincle** [comment]
另外不知道您是否还在继续研究这个课题，CartoonGan的生成效果确实很棒，但是网络的确偏大，处理速度较慢，不知是否有在优化上有什么进展？

**Comment by FlyingGoblin** [comment]
大概是我上传的时间比较晚吧，毕竟最开始并没有整理代码的打算来着（趴
我现在毕业工作了，主要工作并没有在继续这个研究，只是工作间隙还会做一点实验试试。目前没有网缩小网络方向努力，将来有进展可以讨论一下：）

**Comment by wincle** [comment]
嗯，感觉这个工作还是很有趣而且蛮有价值的，如果能够在手机上实时运行是很有亮点的东西，我目前工作是在做风格迁移这块内容，以后需要多多指教了。

**Comment by FlyingGoblin** [comment]
嘻嘻，我之前也是觉得有趣才搞的。指教不敢当，多多讨论多多讨论：）

**Comment by TWLemontree** [comment]
我也有这种疑惑，各种通过 VGG 来进行风格转换的 GAN 网络，论文里都是写的权重为 10，但是我跑的代码里，权重为 10 根本就没作用！

**Comment by TWLemontree** [comment]
我在 models/cartoon_gan_model.lua 中找到这样一段代码：
`
function CartoonGANModel:fGx_basic(x, netG_source, netD_source, real_source, real_target, fake_target,
                                   gradParametersG_source, opt)
  util.BiasZero(netD_source)
  util.BiasZero(netG_source)
  gradParametersG_source:zero()
  -- GAN loss
  local output = netD_source.output -- [hack] forward was already executed in fDx, so save computation netD_source:forward(fake_B) ---
  local errGAN = self.criterionGAN:forward(output, self.real_label)
  local df_do = self.criterionGAN:backward(output, self.real_label)
  local df_d_GAN = netD_source:updateGradInput(fake_target, df_do) ---:narrow(2,fake_AB:size(2)-output_nc+1, output_nc)
  netG_source:forward(real_source)
  netG_source:backward(real_source, df_d_GAN)

  -- content loss
  local errContent, df_d_content = 0, 0
  errContent, df_d_content = content.lossUpdate(self.criterionContent, real_source, fake_target, self.contentFunc, opt.content_loss, self.lambda)
  errContent = errContent/self.lambda
  netG_source:forward(real_source)
  netG_source:backward(real_source, df_d_content)

  return gradParametersG_source, errGAN, errContent
end
`

其中：
`
errContent, df_d_content = content.lossUpdate(self.criterionContent, real_source, fake_target, self.contentFunc, opt.content_loss, self.lambda)  
errContent = errContent/self.lambda
`
这个 content_loss / lambda ，不就相当于 adv_loss : content_loss = 10 : 1 了吗otz

qwq

大佬，快救救孩子吧，毕不了业啦qwq

**Comment by TWLemontree** [comment]
大佬，你的 adv_loss 的确没有乘以 10，但是架不住你的 content_loss 除以 10 鸭！

**Comment by TWLemontree** [comment]
我在util/content_loss.lua里
看到了：
`function content.lossUpdate(criterionContent, real_source, fake_target, contentFunc, loss_type, weight)
  if loss_type == 'none' then
    local errCont = 0.0
    local df_d_content = torch.zeros(fake_target:size())
    return errCont, df_d_content
  elseif loss_type == 'pixel' then
    local errCont = criterionContent:forward(fake_target, real_source) * weight
    local df_do_content = criterionContent:backward(fake_target, real_source)*weight
    print(df_do_content:sum())
    --os.exit()
    return errCont, df_do_content
  elseif loss_type == 'vgg' then
    local f_real = contentFunc:forward(real_source):clone()
    local f_fake = contentFunc:forward(fake_target):clone()
    local errCont = criterionContent:forward(f_fake, f_real) * weight
    local df_do_tmp = criterionContent:backward(f_fake, f_real) * weight
    --print(f_fake:size())
    --print(df_do_tmp:size())
    local df_do_content = contentFunc:updateGradInput(fake_target, df_do_tmp)--:mul(weight)
    --print(df_do_content:size())
    --os.exit()
    --print(df_do_content:sum())
    --os.exit()
    --print(df_do_content:sub(1,1,2,2,3,3))
    --os.exit()
    return errCont, df_do_content
  else error("unsupported content loss")
  end
end
`

其实相当于，您的 content_loss，先在 content.lossUpdate() 里乘了一个 lambda，然后又除以了 lambda，其实 adv_loss : content_loss = 1 : 1

otz

**Comment by FlyingGoblin** [comment]
> 我在 models/cartoon_gan_model.lua 中找到这样一段代码：
> `
> function CartoonGANModel:fGx_basic(x, netG_source, netD_source, real_source, real_target, fake_target,
> gradParametersG_source, opt)
> util.BiasZero(netD_source)
> util.BiasZero(netG_source)
> gradParametersG_source:zero()
> -- GAN loss
> local output = netD_source.output -- [hack] forward was already executed in fDx, so save computation netD_source:forward(fake_B) ---
> local errGAN = self.criterionGAN:forward(output, self.real_label)
> local df_do = self.criterionGAN:backward(output, self.real_label)
> local df_d_GAN = netD_source:updateGradInput(fake_target, df_do) ---:narrow(2,fake_AB:size(2)-output_nc+1, output_nc)
> netG_source:forward(real_source)
> netG_source:backward(real_source, df_d_GAN)
> 
> -- content loss
> local errContent, df_d_content = 0, 0
> errContent, df_d_content = content.lossUpdate(self.criterionContent, real_source, fake_target, self.contentFunc, opt.content_loss, self.lambda)
> errContent = errContent/self.lambda
> netG_source:forward(real_source)
> netG_source:backward(real_source, df_d_content)
> 
> return gradParametersG_source, errGAN, errContent
> end
> `
> 
> 其中：
> `errContent, df_d_content = content.lossUpdate(self.criterionContent, real_source, fake_target, self.contentFunc, opt.content_loss, self.lambda) errContent = errContent/self.lambda`
> 这个 content_loss / lambda ，不就相当于 adv_loss : content_loss = 10 : 1 了吗otz
> 
> qwq
> 
> 大佬，快救救孩子吧，毕不了业啦qwq

抱歉刚看到，我没有太看懂你的点。。。大概是说我代码写的有些问题？不过代码我也不太记得了。。。时间太久了。。。囧。。。你如果还很着急的话可以把你的微信号邮件给我，微信交流快一些，现在我很少查看GitHub了

**Comment by FlyingGoblin** [comment]
> 大佬，你的 adv_loss 的确没有乘以 10，但是架不住你的 content_loss 除以 10 鸭！

我印象中除一下是为了上显示好看，但是backward的时候权重的比例应该是没有问题的

**Comment by TWLemontree** [comment]
> > 大佬，你的 adv_loss 的确没有乘以 10，但是架不住你的 content_loss 除以 10 鸭！
> 
> 我印象中除一下是为了上显示好看，但是backward的时候权重的比例应该是没有问题的

谢谢大佬的回复！我当初遇到的问题其实和该issue的提问者遇到的问题一样，那就是风格化不明显，我自己复现的代码甚至根本就不起作用。最近我看到了另一篇文章AnimeGAN，他将对抗损失的权重设置为300……然后我猜测，可能是因为内容损失的值太大了导致的风格化不起作用。总之还是谢谢大佬啦otz

**Comment by FlyingGoblin** [comment]
> > > 大佬，你的 adv_loss 的确没有乘以 10，但是架不住你的 content_loss 除以 10 鸭！
> > 
> > 
> > 我印象中除一下是为了上显示好看，但是backward的时候权重的比例应该是没有问题的
> 
> 谢谢大佬的回复！我当初遇到的问题其实和该issue的提问者遇到的问题一样，那就是风格化不明显，我自己复现的代码甚至根本就不起作用。最近我看到了另一篇文章AnimeGAN，他将对抗损失的权重设置为300……然后我猜测，可能是因为内容损失的值太大了导致的风格化不起作用。总之还是谢谢大佬啦otz

你解决了你的问题就好：）

**Comment by vinesmsuic** [comment]
> content loss的权重确实不能太大，否则就会出现风格没有办法迁移的问题。 ![image](https://user-images.githubusercontent.com/8320622/46124573-94408600-c257-11e8-91f9-3ca7b71c2fd6.png) 其实我在论文提交时的补充材料里讨论过权重的影响，但是组委会好像后来并没有上传补充材料┓( ´∀` )┏

may i know where can we access the supplementary material?
